src/prebid.js

```diff
@@ -390,63 +390,75 @@ $$PREBID_GLOBAL$$.renderAd = function (doc, id, options) {
     try {
       // lookup ad by ad Id
       const bid = auctionManager.findBidByAdId(id);
+
       if (bid) {
-        // replace macros according to openRTB with price paid = bid.cpm
-        bid.ad = utils.replaceAuctionPrice(bid.ad, bid.cpm);
-        bid.adUrl = utils.replaceAuctionPrice(bid.adUrl, bid.cpm);
-
-        // replacing clickthrough if submitted
-        if (options && options.clickThrough) {
-          const { clickThrough } = options;
-          bid.ad = utils.replaceClickThrough(bid.ad, clickThrough);
-          bid.adUrl = utils.replaceClickThrough(bid.adUrl, clickThrough);
+        let shouldRender = true;
+        if (bid && bid.status === CONSTANTS.BID_STATUS.RENDERED) {
+          utils.logWarn(`Ad id ${bid.adId} has been rendered before`);
+          events.emit(STALE_RENDER, bid);
+          if (utils.deepAccess(config.getConfig('auctionOptions'), 'suppressStaleRender')) {
+            shouldRender = false;
+          }
         }
 
-        // save winning bids
-        auctionManager.addWinningBid(bid);
-
-        // emit 'bid won' event here
-        events.emit(BID_WON, bid);
-
-        const { height, width, ad, mediaType, adUrl, renderer } = bid;
-
-        const creativeComment = document.createComment(`Creative ${bid.creativeId} served by ${bid.bidder} Prebid.js Header Bidding`);
-        utils.insertElement(creativeComment, doc, 'body');
-
-        if (isRendererRequired(renderer)) {
-          executeRenderer(renderer, bid);
-        } else if ((doc === document && !utils.inIframe()) || mediaType === 'video') {
-          const message = `Error trying to write ad. Ad render call ad id ${id} was prevented from writing to the main document.`;
-          emitAdRenderFail({ reason: PREVENT_WRITING_ON_MAIN_DOCUMENT, message, bid, id });
-        } else if (ad) {
-          // will check if browser is firefox and below version 67, if so execute special doc.open()
-          // for details see: https://github.com/prebid/Prebid.js/pull/3524
-          // TODO remove this browser specific code at later date (when Firefox < 67 usage is mostly gone)
-          if (navigator.userAgent && navigator.userAgent.toLowerCase().indexOf('firefox/') > -1) {
-            const firefoxVerRegx = /firefox\/([\d\.]+)/;
-            let firefoxVer = navigator.userAgent.toLowerCase().match(firefoxVerRegx)[1]; // grabs the text in the 1st matching group
-            if (firefoxVer && parseInt(firefoxVer, 10) < 67) {
-              doc.open('text/html', 'replace');
+        if (shouldRender) {
```

Comment:
Do you think for the case in which shouldRender = false there should be an emitRenderFail with a new STALE_RENDER value so that the adRenderFailed event also is triggered?

Reply:
Good question.  It's a possibility that can be added later.  Since there is already a new event, it seems clearer to keep them distinct.

Reply:
Got it. The reason I had asked is that the new event fires regardless if the suppression is enabled, meaning it is not exactly indicative of a failed render. But if it is not a requirement or there does not seem to be a potential issue for pubs leveraging failed render data, then the code looks good to me